### PR TITLE
Display and send votes in polls

### PR DESCRIPTION
**Pull request by andybalaam**
This is the first draft of votes being sent and displayed in polls.

image

Later pull requests will fix some issues:
* a limit on the number of characters in questions in answers
* overlap of option text with the vote count in some screen sizes
* error messages and waiting spinners when actions are happening

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

Preview: https://6197752cb934fbbc06eaaa12--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

**Comment by andybalaam** [comment]
@SimonBrandner we're labelling polls stuff as T-Task for now, so it doesn't appear in the release notes until it's ready to use.

**Comment by SimonBrandner** [comment]
> @SimonBrandner we're labelling polls stuff as T-Task for now, so it doesn't appear in the release notes until it's ready to use.

Ah, sorry for misusing the label, thanks for letting me know

**Comment by andybalaam** [comment]
> Ah, sorry for misusing the label, thanks for letting me know

Actually I think it's probably me misusing the labels, but I guess not calling it a feature when it's hidden behind a labs flag is sort-of legit.

**Comment by Half-Shot** [comment]
I wholeheartedly disagree with the verdict of this poll, but it's a lovely render!

**Comment by andybalaam** [comment]
> I wholeheartedly disagree with the verdict of this poll, but it's a lovely render!

I actually had to re-do the screenshot because I accidentally got the wrong answer first time.

**Review by gaelledel — APPROVED** [review]
Brill :)

File: src/components/views/messages/MPollBody.tsx [hunk 1]
```diff
@@ -18,48 +18,174 @@ import React from 'react';
...
+
+    addListeners(pollRelations?: Relations) {
+        if (pollRelations) {
+            pollRelations.on("Relations.add", this.onRelationsChange);
+            pollRelations.on("Relations.remove", this.onRelationsChange);
+            pollRelations.on("Relations.redaction", this.onRelationsChange);
+        }
+    }
+
+    removeListeners(pollRelations?: Relations) {
```
Comment (turt2live) [inline src/components/views/messages/MPollBody.tsx:75]: explicit member access please (public/private)

File: src/components/views/messages/MPollBody.tsx [hunk 2]
```diff
@@ -18,48 +18,174 @@ import React from 'react';
...
+        this.addListeners(this.state.pollRelations);
+        this.props.mxEvent.on("Event.relationsCreated", this.onPollRelationsCreated);
+    }
+
+    componentWillUnmount() {
+        this.props.mxEvent.off("Event.relationsCreated", this.onPollRelationsCreated);
+        this.removeListeners(this.state.pollRelations);
+    }
+
+    addListeners(pollRelations?: Relations) {
```
Comment (turt2live) [inline src/components/views/messages/MPollBody.tsx:67]: explicit member access please (public/private)

File: src/components/views/messages/MPollBody.tsx [hunk 3]
```diff
@@ -18,48 +18,174 @@ import React from 'react';
...
+            pollRelations.off("Relations.add", this.onRelationsChange);
+            pollRelations.off("Relations.remove", this.onRelationsChange);
+            pollRelations.off("Relations.redaction", this.onRelationsChange);
+        }
     }
 
+    private onPollRelationsCreated = (relationType: string, eventType: string) => {
+        if (
+            relationType === "m.reference" &&
+            eventType === POLL_RESPONSE_EVENT_TYPE.name
```
Comment (turt2live) [inline src/components/views/messages/MPollBody.tsx:86]:
```suggestion
            POLL_RESPONSE_EVENT_TYPE.matches(eventType)
```

File: src/components/views/messages/MPollBody.tsx [hunk 4]
```diff
@@ -18,48 +18,174 @@ import React from 'react';
...
+        return sum;
+    }
+
     render() {
-        const pollStart: IPollContent =
-            this.props.mxEvent.getContent()[POLL_START_EVENT_TYPE.name];
+        const pollStart: IPollContent = this.props.mxEvent.getContent();
+        const pollInfo = pollStart[POLL_START_EVENT_TYPE.name];
+
+        if (pollInfo.answers.length < 1 || pollInfo.answers.length> 20) {
```
Comment (turt2live) [inline src/components/views/messages/MPollBody.tsx:169]:
```suggestion
        if (pollInfo.answers.length < 1 || pollInfo.answers.length > 20) {
```

File: src/components/views/messages/MPollBody.tsx [hunk 5]
```diff
@@ -72,22 +198,122 @@ export default class MPollBody extends React.Component<IBodyProps, IState> {
...
+    ts: number;
+    sender: string;
+    answers: Array<string>;
+
+    constructor(ts: number, sender: string, answers: string[]) {
+        this.ts = ts;
+        this.sender = sender;
+        this.answers = answers;
+    }
+}
```
Comment (turt2live) [inline src/components/views/messages/MPollBody.tsx:234]:
```suggestion
export class UserVote {
    constructor(public readonly ts: number, public readonly sender: string, public readonly answers: string[]) {
    }
}
```

Comment (turt2live) [reply src/components/views/messages/MPollBody.tsx:234]: alternatively:

```typescript
export type UserVote = {
    ts: number;
    sender: string;
    answers: string[];
};

// Example usage: 
function userResponseFromPollResponseEvent(event: MatrixEvent): UserVote {
    // ...
    return { ts: event.getTs(), sender: event.getSender(), answers };
}
```

File: src/components/views/messages/MPollBody.tsx [hunk 6]
```diff
@@ -72,22 +198,122 @@ export default class MPollBody extends React.Component<IBodyProps, IState> {
...
+
+    return new UserVote(
+        event.getTs(),
+        event.getSender(),
+        answers,
+    );
+}
+
+export function allVotes(pollRelations: Relations): Array<UserVote> {
+    function isPollResponse(responseEvent: MatrixEvent) {
```
Comment (turt2live) [inline src/components/views/messages/MPollBody.tsx:248]:
```suggestion
    function isPollResponse(responseEvent: MatrixEvent): boolean {
```

File: src/components/views/messages/MPollBody.tsx [hunk 7]
```diff
@@ -72,22 +198,122 @@ export default class MPollBody extends React.Component<IBodyProps, IState> {
...
+        return (
+            responseEvent.getType() === POLL_RESPONSE_EVENT_TYPE.name &&
+            responseEvent.getContent().hasOwnProperty(POLL_RESPONSE_EVENT_TYPE.name)
+        );
+    }
+
+    if (pollRelations) {
+        return pollRelations.getRelations().filter(isPollResponse).map(
+            userResponseFromPollResponseEvent);
```
Comment (turt2live) [inline src/components/views/messages/MPollBody.tsx:257]: suggestion for clarity:
```suggestion
        return pollRelations.getRelations()
            .filter(isPollResponse)
            .map(userResponseFromPollResponseEvent);
```

**Review by turt2live — CHANGES_REQUESTED** [review]
Overall looks fine - just a few human linter things that we really should put into the lint rules.

File: src/components/views/messages/MPollBody.tsx (hunk 2 above)
Comment (andybalaam) [reply src/components/views/messages/MPollBody.tsx]: I added a commit for this, thanks.

**Comment by andybalaam** [comment]
I have completed all the requested changes, thanks.

**Comment by andybalaam** [comment]
@turt2live this is ready for you to re-review.

**Review by turt2live — APPROVED** [review]
Thanks!

**Pull request merged**